src/msw/textctrl.cpp

```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
 
 void wxTextCtrl::MSWUpdateFontOnDPIChange(const wxSize& newDPI)
 {
-    // Don't do anything for the rich edit controls, they (somehow?) update
-    // their appearance on their own and changing their HFONT, as the base
-    // class version does, would reset all the styles used by them when the DPI
-    // changes, which is unwanted.
+    // Don't use MSWUpdateFontOnDPIChange for the rich edit controls, they
+    // (somehow?) update their appearance on their own and changing their
+    // HFONT, as the base class version does, would reset all the styles used
+    // by them when the DPI changes, which is unwanted.
     if ( !IsRich() )
+    {
         wxTextCtrlBase::MSWUpdateFontOnDPIChange(newDPI);
+    }
+    // If the rich control is created on a screen with non-system DPI, an
+    // initial scaling is applied. This needs to be reset after the first DPI
+    // change. Reapply the styles to the text.
+    else if ( m_richBaseDPI != m_richUsedDPI )
+    {
+        wxEventBlocker block(this, wxEVT_TEXT);
+
+        wxSize baseDPI = m_richBaseDPI;
+
+        wxTextAttr attr;
+        wxTextAttr commonAttr;
+        int rangeStart = -1;
+        int rangeTotal = GetLastPosition();
+
+        for ( int i = 0; i <= rangeTotal; ++i )
+        {
+            // Get style requires the old baseDPI to determine the actually
+            // used font size
+            m_richBaseDPI = baseDPI;
+            GetStyle(i, attr);
+
+
+            if ( attr == commonAttr && i != rangeTotal )
+                continue;
+
+            if ( rangeStart >= 0 || i == rangeTotal )
+            {
+                // update the baseDPI so no font size change is applied
+                m_richBaseDPI = m_richUsedDPI;
+                SetStyle(rangeStart, i, commonAttr);
+            }
+
+            rangeStart = i;
+            commonAttr = attr;
+        }
+    }
```

Comment:
What's the value of `m_richBaseDPI` at the end here? It looks like it could have changed, shouldn't we restore it to its original value?

Reply:
It ends with the value of `m_richUsedDPI`. So any subsequent DPI changes, this code is not executed anymore.

Reply:
OK, thanks, I see now that we always execute `SetStyle()` call due to `i == rangeTotal` check. This is a bit tortuous though, I think doing
```cpp
m_richBaseDPI = baseDPI;
GetStyle(i, attr);
m_richBaseDPI = m_richUsedDPI;
```
and not touching `m_richBaseDPI` anywhere else would make it more clear that we just use the base DPI temporarily for `GetStyle()` call duration and that the base DPI always is the same as used DPI at the end.